### PR TITLE
SMC sampling update; cumulative logits, changed normalization

### PR DESCRIPTION
**Pull request by culpritgene**

### CODE REVIEW

File: refactored/model.py [hunk 1]
```diff
@@ -43,7 +43,7 @@ def get_base_model_logits_for_sequence(self, sequence):
         input_ids = torch.tensor(sequence, device=self.device).unsqueeze(0)
         with torch.no_grad():
             outputs = self.base_model(input_ids=input_ids, return_dict=True)
-            return outputs.logits[0, -1, :]
+            return outputs.logits.squeeze(0)[:, -1, :]  # NOTE: now may fail for single particle run
```
Comment (culpritgene) [inline refactored/model.py:46]: for batch_size > 1 previous code may truncate the batch
though the output seems to be a 4d tensor [1, batch_size, seq_length, dim]
Still may fail on other models, even if they are still from transformers lib

File: refactored/smc_sampling.py [hunk 2]
```diff
@@ -1,137 +1,250 @@
...
+        if ess < num_particles / 2 and (t < new_tokens_count - 1):
+            logger.info(f"  Step {t+1}: ESS {ess.item():.2f} < {num_particles / 2}. Resampling.")
+            (
+                resampled_particle_indices,
+                (
+                    curr_particle_sequences,
+                    curr_log_p_cumulative,
+                    curr_log_psi_cumulative,
+                )
+            ) = resample_particles_and_state(
```
Comment (culpritgene) [inline refactored/smc_sampling.py:174]: Simplified the method reducing number of carried over variables

File: refactored/smc_sampling.py [hunk 3]
```diff
@@ -1,137 +1,250 @@
...
+                curr_log_weights,
+                curr_log_p_cumulative,
+                curr_log_psi_cumulative,
+            )
+            # Important Step from paper: reset log_weights to zeros after resampling
+            curr_log_weights = torch.zeros(num_particles, device=device)
+        elif logger:
+            logger.debug(f"  Step {t+1}: ESS {ess.item():.2f} >= {num_particles / 2}. No resampling.")
+
+        if record_log_psi_incrementals:
```
Comment (culpritgene) [inline refactored/smc_sampling.py:185]: Currently we only support accumulation of record_log_psi_incrementals, code will fail if the flag is turned off.
